From 0b764b0767da9edb4bf4dd8a600bf36a374667e9 Mon Sep 17 00:00:00 2001
From: Thang Tran <thuutran@amperecomputing.com>
Date: Thu, 24 Nov 2022 10:45:47 +0700
Subject: [PATCH] Control.Power.Limit: Add power limit interface

Based on the chapter 6.6 - Power management of DCMI specification, this
interface provides the properties as below:
  1. Active: Enable/Disable the power limit
  2. GroupExtensionId: Group Extension ID (0xDC)
  3. ExceptionAction: Exception Action
  4. PowerLimit: Power limit requested in watts
  5. CorrectionTime: Correction time in milliseconds
  6. SamplingPeriod: Sampling period in seconds

Those properties are used to monitor total power consumption, depending
on the user's configuration, BMC can turn off the power/write to SEL log
or do nothing when the total power consumption is greater than the power
limit.

Signed-off-by: Thang Tran <thuutran@amperecomputing.com>
---
 .../Control/Power/Limit/meson.build           |  15 +++
 .../openbmc_project/Control/Power/meson.build |  14 ++
 .../Control/Power/Limit.interface.yaml        | 124 ++++++++++++++++++
 3 files changed, 153 insertions(+)
 create mode 100644 gen/xyz/openbmc_project/Control/Power/Limit/meson.build
 create mode 100644 yaml/xyz/openbmc_project/Control/Power/Limit.interface.yaml

diff --git a/gen/xyz/openbmc_project/Control/Power/Limit/meson.build b/gen/xyz/openbmc_project/Control/Power/Limit/meson.build
new file mode 100644
index 0000000..6887bf2
--- /dev/null
+++ b/gen/xyz/openbmc_project/Control/Power/Limit/meson.build
@@ -0,0 +1,15 @@
+# Generated file; do not modify.
+generated_sources += custom_target(
+    'xyz/openbmc_project/Control/Power/Limit__cpp'.underscorify(),
+    input: [ '../../../../../../yaml/xyz/openbmc_project/Control/Power/Limit.interface.yaml',  ],
+    output: [ 'server.cpp', 'server.hpp', 'client.hpp',  ],
+    depend_files: sdbusplusplus_depfiles,
+    command: [
+        sdbuspp_gen_meson_prog, '--command', 'cpp',
+        '--output', meson.current_build_dir(),
+        '--tool', sdbusplusplus_prog,
+        '--directory', meson.current_source_dir() / '../../../../../../yaml',
+        'xyz/openbmc_project/Control/Power/Limit',
+    ],
+)
+
diff --git a/gen/xyz/openbmc_project/Control/Power/meson.build b/gen/xyz/openbmc_project/Control/Power/meson.build
index 8ea227b..ee013ac 100644
--- a/gen/xyz/openbmc_project/Control/Power/meson.build
+++ b/gen/xyz/openbmc_project/Control/Power/meson.build
@@ -74,3 +74,17 @@ generated_others += custom_target(
     ],
 )
 
+subdir('Limit')
+generated_others += custom_target(
+    'xyz/openbmc_project/Control/Power/Limit__markdown'.underscorify(),
+    input: [ '../../../../../yaml/xyz/openbmc_project/Control/Power/Limit.interface.yaml',  ],
+    output: [ 'Limit.md' ],
+    depend_files: sdbusplusplus_depfiles,
+    command: [
+        sdbuspp_gen_meson_prog, '--command', 'markdown',
+        '--output', meson.current_build_dir(),
+        '--tool', sdbusplusplus_prog,
+        '--directory', meson.current_source_dir() / '../../../../../yaml',
+        'xyz/openbmc_project/Control/Power/Limit',
+    ],
+)
\ No newline at end of file
diff --git a/yaml/xyz/openbmc_project/Control/Power/Limit.interface.yaml b/yaml/xyz/openbmc_project/Control/Power/Limit.interface.yaml
new file mode 100644
index 0000000..3e83446
--- /dev/null
+++ b/yaml/xyz/openbmc_project/Control/Power/Limit.interface.yaml
@@ -0,0 +1,124 @@
+description: >
+    The power limit parameters on the system which are define in the DCMI
+    specification
+
+properties:
+    - name: Active
+      type: boolean
+      default: false
+      description: >
+          This property indicate the power limit feature is enable or not.
+
+    - name: GroupExtensionId
+      type: byte
+      default: 220
+      flags:
+          - readonly
+      description: >
+          The Group Extension Identification, it always is 0xDC.
+
+    - name: ExceptionAction
+      type: enum[self.ExceptionActions]
+      description: >
+          Exception Actions, taken if the Power Limit is exceeded and cannot be
+          controlled within the Correction time limit
+
+    - name: PowerLimit
+      type: uint16
+      description: >
+          This property contain the maximum power consumption of the system.
+
+    - name: CorrectionTime
+      type: uint32
+      description: >
+          Maximum time taken to limit the power after the platform power has 
+          reached the power limit before the Exception Action will be taken. 
+          The Exception Action shall be taken if the system power usage 
+          constantly exceeds the specified power limit for more than the 
+          Correction Time Limit interval. The Correction Time Limit timeout 
+          automatically restarts if the system power meets or drops below the 
+          Power Limit.
+
+    - name: SamplingPeriod
+      type: uint16
+      default: 1
+      description: >
+          Management application Statistics Sampling period in seconds
+
+enumerations:
+    - name: ExceptionActions
+      description: >
+          Exception Actions, taken if the Power Limit is exceeded and cannot be
+          controlled within the Correction Time Limit
+      values:
+          - name: NoAction
+            description: >
+                No Action
+
+          - name: HardPowerOff
+            description: >
+                Hard Power Off system and log event to SEL
+
+          - name: OEM02
+            decription: >
+                OEM (02) defined action
+
+          - name: OEM03
+            decription: >
+                OEM (03) defined action
+
+          - name: OEM04
+            decription: >
+                OEM (04) defined action
+
+          - name: OEM05
+            decription: >
+                OEM (05) defined action
+
+          - name: OEM06
+            decription: >
+                OEM (06) defined action
+
+          - name: OEM07
+            decription: >
+                OEM (07) defined action
+
+          - name: OEM08
+            decription: >
+                OEM (08) defined action
+
+          - name: OEM09
+            decription: >
+                OEM (09) defined action
+
+          - name: OEM0A
+            decription: >
+                OEM (0A) defined action
+
+          - name: OEM0B
+            decription: >
+                OEM (0B) defined action
+
+          - name: OEM0C
+            decription: >
+                OEM (0C) defined action
+
+          - name: OEM0D
+            decription: >
+                OEM (0D) defined action
+
+          - name: OEM0E
+            decription: >
+                OEM (0E) defined action
+
+          - name: OEM0F
+            decription: >
+                OEM (0F) defined action
+
+          - name: OEM10
+            decription: >
+                OEM (10) defined action
+
+          - name: SELLog
+            decription: >
+                SEL log only
-- 
2.30.0

